src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookConsumer.java

```diff
@@ -26,90 +28,144 @@
+            }
         }
-        return Collections.emptySet();
+        return false;
     }
 
-    private static boolean matchingRef(Set<String> refChanges, GitSCM gitSCM) {
-        List<BranchSpec> branchSpecs = gitSCM.getBranches();
-        return refChanges
-                .stream()
-                .anyMatch(
-                        ref ->
-                                branchSpecs
-                                        .stream()
-                                        .anyMatch(branchSpec -> branchSpec.matches(ref)));
+    private static boolean hasMatchingRepository(RefChangedDetails refChangedDetails,
+                                                 BitbucketSCM scm) {
+        return scm.getRepositories().stream()
+                .anyMatch(scmRepo -> matchingRepo(refChangedDetails.getRepository(), scmRepo));
+    }
+
+    private static boolean hasMatchingRepository(RefChangedDetails refChangedDetails,
+                                                 GitSCM scm) {
+        return scm.getRepositories().stream()
+                .anyMatch(scmRepo -> matchingRepo(refChangedDetails.getCloneLinks(), scmRepo));
+    }
+
+    private static boolean matchingRepo(BitbucketRepository repository, BitbucketSCMRepository scmRepo) {
+        return scmRepo.getProjectKey().equals(repository.getProject().getKey())
+                && scmRepo.getRepositorySlug().equals(repository.getSlug());
     }
 
     private static boolean matchingRepo(Set<String> cloneLinks, RemoteConfig repo) {
         return repo.getURIs().stream().anyMatch(uri -> cloneLinks.contains(uri.toString()));
     }
 
+    private static Optional<TriggerDetails> toTriggerDetails(ParameterizedJobMixIn.ParameterizedJob job) {
+        BitbucketWebhookTriggerImpl trigger = triggerFrom(job);
+        if (trigger != null) {
+            return Optional.of(new TriggerDetails(job, trigger));
+        }
+        return Optional.empty();
+    }
+
     @Nullable
-    private static BitbucketWebhookTriggerImpl triggerFrom(
-            ParameterizedJobMixIn.ParameterizedJob item) {
-        Map<TriggerDescriptor, Trigger<?>> triggers = item.getTriggers();
+    private static BitbucketWebhookTriggerImpl triggerFrom(ParameterizedJobMixIn.ParameterizedJob job) {
+        Map<TriggerDescriptor, Trigger<?>> triggers = job.getTriggers();
         for (Trigger candidate : triggers.values()) {
             if (candidate instanceof BitbucketWebhookTriggerImpl) {
                 return (BitbucketWebhookTriggerImpl) candidate;
             }
         }
         return null;
     }
+
+    private static Collection<? extends SCM> getScms(ParameterizedJobMixIn.ParameterizedJob job) {
+        SCMTriggerItem triggerItem = SCMTriggerItem.SCMTriggerItems.asSCMTriggerItem(job);
+        if (triggerItem != null) {
+            return triggerItem.getSCMs();
+        }
+        return Collections.emptySet();
+    }
+
+    private static class TriggerDetails {
```

Comment:
This class and `RefChangedDetails` can both be final and have private constructors